services/route_manager/src/main/java/com/futurewei/alcor/route/service/Impl/NeutronRouterServiceImpl.java

```diff
@@ -0,0 +1,484 @@
+            // get subnet by port id
+            SubnetsWebJson subnetsWebJson = this.routerToSubnetService.getSubnetsByPortId(projectid, portId);
+            if (subnetsWebJson == null) {
+                return new RouterInterfaceResponse();
+            }
+            ArrayList<SubnetEntity> subnets = subnetsWebJson.getSubnets();
+            if (subnets.size() == 0) {
+                return new RouterInterfaceResponse();
+            }
+            if (subnets.size() != 1) {
+                throw new SubnetNotBindUniquePortId();
+            }
+            subnet = subnets.get(0);
+            subnetid = subnet.getId();
+        }else if (portId == null && subnetId != null) {
+            // get subnet by subnet id
+            SubnetWebJson subnetWebJson = this.routerToSubnetService.getSubnet(projectid, subnetId);
+            subnet = subnetWebJson.getSubnet();
+            if (subnet == null) {
+                logger.warn("can not find subnet by subnet id :" + subnetId);
+                return new RouterInterfaceResponse();
+            }
+            subnetid = subnetId;
+        }else {
+            return new RouterInterfaceResponse();
+        }
+
+        // check if the router or the subnet and port do not exist or are not visible
+        Router router = this.routerDatabaseService.getByRouterId(routerId);
+        if (router == null) {
+            throw new RouterOrSubnetAndPortNotExistOrNotVisible();
+        }
+
+        // check if you try to delete the router interface for subnets that are used by one or more route
+        projectId = subnet.getProjectId();
+        attachedRouterId = subnet.getAttachedRouterId();
+        String gatewayIp = subnet.getGatewayIp();
+        if (gatewayIp != null) {
+            RouteTable routeTable = router.getRouteTable();
+            if (routeTable == null) {
+                throw new RouterTableNotExist();
+            }
+
+            List<RouteEntry> routeEntities = routeTable.getRouteEntities();
+            for (RouteEntry routeEntry : routeEntities) {
+                String nextHop = routeEntry.getNexthop();
+                if (gatewayIp.equals(nextHop)) {
+                    throw new RouterInterfaceAreUsedByRoutes();
+                }
+            }
+
+        }
+
+        // remove interface
+        subnet.setAttachedRouterId("");
+
+        List<String> ports = router.getPorts();
+        if (ports == null){
+            return new RouterInterfaceResponse();
+        }
+
+        if (ports.contains(portId)) {
+            ports.remove(portId);
+        }
+        router.setPorts(ports);
+        this.routerDatabaseService.addRouter(router);
+
+        // update device_id and device_owner
+        PortEntity portEntity = new PortEntity();
+        portEntity.setDeviceId(null);
+        portEntity.setDeviceOwner(null);
+        subnet.setPort(portEntity);
+
+        // update subnet
+        this.routerToSubnetService.updateSubnet(projectId, subnetid, subnet);
+
```

Comment:
@kevin-zhonghao We also need to update the port-subnet mapping table in the Route Manager.

Reply:
leave it for next PR